mindmeld/augmentation.py

```diff
@@ -0,0 +1,340 @@
+        )
+
+        MarianTokenizer = _get_module_or_attr("transformers", "MarianTokenizer")
+        MarianMTModel = _get_module_or_attr("transformers", "MarianMTModel")
+
+        en_model_name = "Helsinki-NLP/opus-mt-ROMANCE-en"
+        self.en_tokenizer = MarianTokenizer.from_pretrained(en_model_name)
+        self.en_model = MarianMTModel.from_pretrained(en_model_name)
+        self.en_model.to(self.torch_device)
+
+        target_model_name = "Helsinki-NLP/opus-mt-en-ROMANCE"
+        self.target_tokenizer = MarianTokenizer.from_pretrained(target_model_name)
+        self.target_model = MarianMTModel.from_pretrained(target_model_name).to(
+            self.torch_device
+        )
+
+        # Update default params with user model config
+        self.fwd_params = {
+            "num_beams": 3,
+            "num_return_sequences": 3,
+            "top_k": 0,
+            "temperature": 1.0,
+        }
+
+        self.reverse_params = copy.deepcopy(self.fwd_params)
+
+        self.fwd_params.update(params.get("fwd_params", {}))
+        self.reverse_params.update(params.get("reverse_params", {}))
+        self.batch_size = params.get("batch_size", 8)
+
+    def _translate(self, *, queries, model, tokenizer, **kwargs):
+        """The core translation step for forward and reverse translation.
+
+        Args:
+            template (lambda func): Structure input text to model.
+            queries (list(str)): List of input queries.
+            model: Machine translation model (en-ROMANCE or ROMANCE-en).
+            tokenizer: Language tokenizer for input query text.
+        """
+        all_translated_queries = []
+        for pos in range(0, len(queries), self.batch_size):
+            encoded = tokenizer.prepare_seq2seq_batch(
+                queries[pos : pos + self.batch_size], return_tensors="pt"
+            )
+            for key in encoded:
+                encoded[key] = encoded[key].to(self.torch_device)
+
+            translated = model.generate(**encoded, **kwargs)
+            translated_queries = tokenizer.batch_decode(
+                translated, skip_special_tokens=True
+            )
+            all_translated_queries.extend(translated_queries)
+        return all_translated_queries
+
+    def augment_queries(self, queries):
+        translated_queries = self._translate(
+            queries=queries,
+            model=self.en_model,
+            tokenizer=self.en_tokenizer,
+            **self.fwd_params,
+        )
+
+        template = lambda text: f">>{self.lang}<< {text}"
+        translated_queries = [template(query) for query in set(translated_queries)]
+
+        reverse_translated_queries = self._translate(
+            queries=translated_queries,
+            model=self.target_model,
+            tokenizer=self.target_tokenizer,
+            **self.reverse_params,
+        )
+        augmented_queries = list(set(p.lower() for p in reverse_translated_queries))
```

Comment:
From my experiments with WxA, for some of the shorter queries, it seems to generate sentences with just punctuations sometimes. One way of eliminating them could be to retain queries that have some alpha-numeric characters.

Reply:
Consider something like this - 
```
augmented_queries = list(set([p.lower() for p in reverse_translated_queries if any(c.isalnum() for c in p)]))
```